platform/src/main/java/org/hillview/sketches/Centroids.java

```diff
@@ -0,0 +1,101 @@
+package org.hillview.sketches;
+
+import org.hillview.dataset.api.IJson;
+import org.hillview.table.api.IMembershipSet;
+import org.hillview.table.api.ITable;
+import org.hillview.utils.BlasConversions;
+import org.jblas.DoubleMatrix;
+import org.jblas.ranges.AllRange;
+import org.jblas.ranges.PointRange;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+/**
+ * This class represents a set of centroids of a partitioning of a table. The number of centroids should not be too
+ * large, as they are shipped over the network.
+ */
+public class Centroids implements IJson {
```

Comment:
You have here a bit too much; when you send the data to the client you don't actually want to serialize everything, only the actual centroids. Maybe you should not implement IJson at all, you just return from the web server the result of the call of computeCentroids.

Reply:
Oh right, so if I understand correctly, objects that have to be passed between backend nodes should implement just `Serializable`, while objects that have to be passed to the client (i.e., browser) should implement `IJson` (which also implements `Serializable`).

Reply:
IJson does not extend Serializable; Serializable means "Serialize the object the way Java knows", while IJson means "Serialize the object as JSON".
Unfortunately the IJson interface is not first-class like Serializable is, so just inheriting from IJson does not make it work automatically; sometimes you may need to do additional work, like registering some serialization/deserialization classes in the IJson file (take a look).